beard/clustering/blocking.py

```diff
@@ -276,12 +269,32 @@ def predict(self, X, blocks=None):
             # Predict on the block, if known
             if b in self.clusterers_:
                 mask = (blocks == b)
-                cluster = self.clusterers_[b]
+                clusterer = self.clusterers_[b]
 
-                pred = np.array(cluster.predict(X[mask]))
+                pred = np.array(clusterer.predict(X[mask]))
                 mask_unknown = (pred == -1)
                 pred[~mask_unknown] += offset
                 labels[mask] = pred
-                offset += np.max(cluster.labels_) + 1
+                offset += np.max(clusterer.labels_) + 1
+
+        return labels
+
+    @property
+    def labels_(self):
+        """Compute the labels assigned to the input data.
+
+        Note that labels are computed on-the-fly.
+        """
+        labels = -np.ones(len(self.blocks_), dtype=np.int)
+        offset = 0
+
+        for b in self.clusterers_:
+            clusterer = self.clusterers_[b]
+
+            pred = np.array(clusterer.labels_)
+            mask_unknown = (pred == -1)
+            pred[~mask_unknown] += offset
+            labels[(self.blocks_ == b)] = pred
+            offset += np.max(clusterer.labels_) + 1
```

Comment:
OK, it's not that bad. Let 's just remember that when we do the same thing one more time, we should factor it out. :+1: